File: frontend/src/domain/types/beaconStatus.js
/**
 * @typedef BeaconStatus
 * @property {number} id
 * @property {string} internalReferenceNumber
 * @property {string} externalReferenceNumber
 * @property {string} ircs
 * @property {string} vesselIdentifier
 * @property {string} vesselName
 * @property {string} vesselStatus
 * @property {string} stage
 * @property {boolean} priority
 * @property {string} malfunctionStartDateTime
 * @property {string | null} malfunctionEndDateTime
 * @property {string} vesselStatusLastModificationDateTime
 */

/**
 * @typedef UpdateBeaconStatus
 * @property {string | null} [vesselStatus]
 * @property {string | null} [stage]
 */

/**
 * @typedef BeaconStatusComment
 * @property {number} id
 * @property {string} comment
 * @property {string} userType
 * @property {string} dateTime
 */

/**
 * @typedef BeaconStatusAction
 * @property {number} id
 * @property {int} beaconStatusId
 * @property {string} propertyName
 * @property {string} previousValue
 * @property {string} nextValue
 * @property {string} dateTime
 */

/**
 * @typedef BeaconStatusCommentInput
 * @property {string} comment
 * @property {string} userType
 */

/**
 * @typedef BeaconStatusWithDetails
 * @property {BeaconStatus} beaconStatus
 * @property {BeaconStatusComment[]} comments
 * @property {BeaconStatusAction[]} actions
 */

/**
 * @typedef VesselBeaconMalfunctionsResumeAndHistory
 * @property {BeaconStatusWithDetails || null} current
 * @property {VesselBeaconMalfunctionsResume} resume
 * @property {BeaconStatusWithDetails[]} history
 * @property {VesselIdentity} vesselIdentity
 */

/**
 * @typedef VesselBeaconMalfunctionsResume
 * @property {number} numberOfBeaconsAtSea
 * @property {number} numberOfBeaconsAtPort
 * @property {string || null} lastBeaconStatusDateTime
 * @property {string || null} lastBeaconStatusVesselStatus
 */

File: backend/src/main/kotlin/fr/gouv/cnsp/monitorfish/domain/entities/beacon_statuses/BeaconStatus.kt
package fr.gouv.cnsp.monitorfish.domain.entities.beacon_statuses

import fr.gouv.cnsp.monitorfish.domain.entities.VesselIdentifier
import fr.gouv.cnsp.monitorfish.domain.entities.last_position.LastPosition
import java.time.ZonedDateTime

data class BeaconStatus(
        val id: Int,
        val internalReferenceNumber: String?,
        val externalReferenceNumber: String?,
        val ircs: String?,
        val vesselIdentifier: VesselIdentifier,
        val vesselName: String,
        val vesselStatus: VesselStatus,
        val stage: Stage,
        val priority: Boolean,
        val malfunctionStartDateTime: ZonedDateTime,
        val malfunctionEndDateTime: ZonedDateTime?,
        val vesselStatusLastModificationDateTime: ZonedDateTime,
        var riskFactor: Double? = null) {
    companion object {
        fun getVesselFromBeaconStatus(beaconStatus: BeaconStatus): (LastPosition) -> Boolean {
            return { lastPosition ->
                when (beaconStatus.vesselIdentifier) {
                    VesselIdentifier.INTERNAL_REFERENCE_NUMBER -> lastPosition.internalReferenceNumber == beaconStatus.internalReferenceNumber
                    VesselIdentifier.IRCS -> lastPosition.ircs == beaconStatus.ircs
                    VesselIdentifier.EXTERNAL_REFERENCE_NUMBER -> lastPosition.externalReferenceNumber == beaconStatus.externalReferenceNumber
                    else -> false
                }
            }
        }
    }
}